palace/utils/geodata.cpp

```diff
@@ -1571,7 +1615,344 @@ std::unique_ptr<mfem::Mesh> LoadMesh(const std::string &path, bool remove_curvat
+      mesh = &new_mesh;
+    }
+  }
+
+  // Convert all element types to hexahedra (currently only tet-to-hex).
+  if (make_hex)
+  {
+    const auto element_types = mesh::CheckElements(*mesh);
+    if (element_types.has_simplices || element_types.has_prisms ||
+        element_types.has_pyramids)
+    {
+      MFEM_VERIFY(!mesh->Nonconforming(),
+                  "Mesh element splitting is not supported for nonconforming meshes!");
+      MFEM_VERIFY(!element_types.has_prisms && !element_types.has_pyramids,
+                  "Splitting mesh elements to hexahedra only supports simplex elements "
+                  "(tetrahedra) for now!");
+      if (preserve_curvature && mesh->GetNodes() && !orig_mesh_attr.Size())
+      {
+        BackUpAttributes(*mesh);
+      }
+      int ne = mesh->GetNE();
+      new_mesh = MeshTetToHex(*mesh);
+      Mpi::Print("Added {:d} elements to the mesh during conversion to hexahedra\n",
+                 new_mesh.GetNE() - ne);
+      mesh = &new_mesh;
+    }
+  }
+
+  // Return if no modifications were made.
+  if (mesh == orig_mesh.get())
+  {
+    return;
+  }
+
+  // The previous splitting functions remove curvature information from the new mesh. So, if
+  // needed, we interpolate it onto the new mesh with GSLIB. The topology of the mesh must
+  // be finalized for this to work (assumes FinalizeTopology has been called on the new
+  // mesh).
+  if (preserve_curvature && orig_mesh->GetNodes())
+  {
+    // Prepare to interpolate the grid function for high-order nodes from the old mesh to
+    // the new one. This first sets up the new mesh as a linear mesh and constructs a
+    // separate high-order grid function for storing the interpolated nodes.
+    new_mesh.EnsureNodes();
+    const mfem::GridFunction *nodes = orig_mesh->GetNodes();
+    const mfem::FiniteElementSpace *fespace = nodes->FESpace();
+    mfem::Ordering::Type ordering = fespace->GetOrdering();
+    int order = fespace->GetMaxElementOrder();
+    int sdim = orig_mesh->SpaceDimension();
+    bool discont =
+        (dynamic_cast<const mfem::L2_FECollection *>(fespace->FEColl()) != nullptr);
+    mfem::FiniteElementSpace new_fespace(&new_mesh, fespace->FEColl(), sdim, ordering);
+    mfem::GridFunction new_nodes(&new_fespace);
+
+    mfem::Array<int> vdofs;
+    mfem::Vector vals, elem_vals, xyz;
+    mfem::DenseMatrix pointmat;
+    int start = 0;
+    for (int e = 0; e < orig_mesh->GetNE(); e++)
+    {
+      // Get the high-order nodes restricted to this parent element, always returned
+      // byNODES.
+      fespace->GetElementVDofs(e, vdofs);
+      nodes->GetSubVector(vdofs, vals);
+
+      // Find all child elements of this parent in the split mesh and restore their correct
+      // original attribute.
+      int attr = new_mesh.GetAttribute(start);
+      int end = start;
+      while (end < new_mesh.GetNE() && new_mesh.GetAttribute(end) == attr)
+      {
+        new_mesh.SetAttribute(end++, orig_mesh_attr[e]);
+      }
```

Comment:
This loop works because the elements are added as `[[new_00, new_01, new_02, ...], [new_10, new_11, new_12,...],...]`, namely all new elements that made up an original element are contiguous. Also recursive, so tet -> hex -> tet will still have this contiguous structure.

Reply:
I'll add a comment for this.